core/src/main/java/org/sonarsource/sonarlint/core/log/MessageFormat.java

```diff
@@ -0,0 +1,490 @@
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 28 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void handleRemainingCharIfAny(final String messagePattern, final int len,
+    final StringBuilder buffer, final int escapeCounter, final int i) {
+    if (i == len - 1) {
+      final char curChar = messagePattern.charAt(i);
+      handleLastChar(buffer, escapeCounter, curChar);
+    }
+  }
+
+  /**
+   * Processes the last unprocessed character and returns the resulting position in the result char array.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 28 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void handleLastChar(final StringBuilder buffer, final int escapeCounter, final char curChar) {
+    if (curChar == ESCAPE_CHAR) {
+      writeUnescapedEscapeChars(escapeCounter + 1, buffer);
+    } else {
+      handleLiteralChar(buffer, escapeCounter, curChar);
+    }
+  }
+
+  /**
+   * Processes a literal char (neither an '\' escape char nor a "{}" delimiter pair) and returns the resulting
+   * position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 16 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void handleLiteralChar(final StringBuilder buffer, final int escapeCounter, final char curChar) {
+    // any other char beside ESCAPE or DELIM_START/STOP-combo
+    // write unescaped escape chars
+    writeUnescapedEscapeChars(escapeCounter, buffer);
+    buffer.append(curChar);
+  }
+
+  /**
+   * Writes "{}" to the specified result array at the specified position and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 18 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeDelimPair(final StringBuilder buffer) {
+    buffer.append(DELIM_START);
+    buffer.append(DELIM_STOP);
+  }
+
+  /**
+   * Returns {@code true} if the specified parameter is odd.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 11 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static boolean isOdd(final int number) {
+    return (number & 1) == 1;
+  }
+
+  /**
+   * Writes a '\' char to the specified result array (starting at the specified position) for each <em>pair</em> of
+   * '\' escape chars encountered in the message format and returns the resulting position.
+   */
+  // Profiling showed this method is important to log4j performance. Modify with care!
+  // 11 bytes (allows immediate JVM inlining: < 35 bytes) LOG4J2-1096
+  private static void writeEscapedEscapeChars(final int escapeCounter, final StringBuilder buffer) {
+    final int escapedEscapes = escapeCounter >> 1; // divide by two
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Move this trailing comment on the previous empty line. ([squid:TrailingCommentCheck](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarlint.core%3Asonarlint-core-parent&pullRequest=165&issues=AWM_D9R25HoPdzsInowW&open=AWM_D9R25HoPdzsInowW)

[ISSUE_KEY]: AWM_D9R25HoPdzsInowW